File: task_scheduling.py
from typing import List
from collections import deque

def task_scheduling(tasks: List[str], requirements: List[List[str]]) -> List[str]:
    # WRITE YOUR BRILLIANT CODE HERE
    def get_node_degree(tasks, requirements):
        degrees = {task:0 for task in tasks}
        for src, des in requirements:
            degrees[des] += 1
        return degrees
    
    def create_graph(tasks, requirements):
        graph = {task:[] for task in tasks}
        for x, y in requirements:
            graph[x] = graph.get(x, []) + [y]
        return graph
        
    res = []
    queue = deque([])
    graph = create_graph(tasks, requirements)
    degrees = get_node_degree(tasks, requirements)
    for node, degree in degrees.items():
        if degree == 0:
            queue.append(node)
            
    while queue:
        node = queue.popleft()
        if degrees[node] == 0:
            res.append(node)
        for child in graph[node]:
            degrees[child] -= 1
            if degrees[child] == 0:
                queue.append(child)
    
    return res


# solution
from collections import deque
from typing import List

def count_parents(graph):
    counts = { node: 0 for node in graph }
    for parent in graph:
        for node in graph[parent]:
            counts[node] += 1
    return counts


def topo_sort(graph):
    res = []
    q = deque()
    counts = count_parents(graph)
    for node in counts:
        if counts[node] == 0:
            q.append(node)
    while len(q) > 0:
        node = q.popleft()
        res.append(node)
        for child in graph[node]:
            counts[child] -= 1
            if counts[child] == 0:
                q.append(child)
    return res if len(graph) == len(res) else None

def task_scheduling(tasks: List[str], requirements: List[List[str]]) -> List[str]:
    graph = {t: [] for t in tasks}
    for a, b in requirements:
        graph[a].append(b)
    return topo_sort(graph)

File: js/task_scheduling.js
function taskScheduling(tasks, requirements) {
    // WRITE YOUR BRILLIANT CODE HERE
    let indegree = new Map();
    let queue = [];
    let res = [];
    let graph = new Map();
    for (const task of tasks) {
        indegree.set(task, 0);
        graph.set(task, []);
    }
    for (const [src, dst] of requirements) {
        indegree.set(dst, indegree.get(dst) + 1);
        graph.get(src).push(dst);
    }
    for (const node of indegree.keys()) {
        if (indegree.get(node) === 0) {
            queue.push(node);
        }
    }
    
    while (queue.length > 0) {
        const node = queue.shift();
        res.push(node);
        for (const neighbor of graph.get(node)) {
            indegree.set(neighbor, indegree.get(neighbor) - 1);
            if (indegree.get(neighbor) === 0) {
                queue.push(neighbor);
            }
        }
    }
    
    return (graph.size === res.length) ? res : null;
}

// algomonster solution
function countParents(graph) {
    const counts = new Map();
    for (const node of graph.keys()) {
        counts.set(node, 0);
    }
    for (const parent of graph.keys()) {
        for (node of graph.get(parent)) {
            counts.set(node, counts.get(node) + 1);
        }
    }
    return counts;
}

function topoSort(graph) {
    const res = [];
    const q = [];
    const counts = countParents(graph);
    for (const node of counts.keys()) {
        if (counts.get(node) == 0) {
            q.push(node);
        }
    }
    while (q.length > 0) {
        const node = q.shift();
        res.push(node);
        for (const child of graph.get(node)) {
            counts.set(child, counts.get(child) - 1);
            if (counts.get(child) == 0) {
                q.push(child);
            }
        }
    }
    return (graph.size === res.length) ? res : null;
}

function taskScheduling(tasks, requirements) {
    const graph = new Map();
    for (let task of tasks) {
        graph.set(task, []);
    }
    for (let req of requirements) {
        graph.get(req[0]).push(req[1]);
    }
    return topoSort(graph);
}